### ISSUE TITLE
<> key and $§ key swapped on v15.0.2

### ISSUE THREAD

**Issue by GeoffGMartin** [pqrs-org/Karabiner-Elements#3984]
- iMac 3.4 GHz Quad Core i5
- Ventura 13.7
- Apple Danish Bluetooth keyboard with Keypad

- After upgrade from Karabiner 14.13.0 to 15.0.2, the '<>' key and the '$§' key swapped behaviours.

- Uninstalled 15.0.2 and downgraded to 14.13.0: problem solved

**Comment by arthur-eudeline** [comment]
Yeah, I think I'm having the same problem as you on an AZERTY Logi MX Keyboard, the "@²" key is swapped with the "<" key as shown on the screenshot below : 
![mx-keys-s-keyboard-top-view-graphite-fra - copie](https://github.com/user-attachments/assets/29792976-7458-43c5-ab01-bac95f0220a5)

**Comment by lsnizek** [comment]
Try changing type from ISO to ANSI under Virtual Keyboard in the settings

**Comment by Chabulga** [comment]
Same issue here on mac OS. 

I swapped the keys manually by cross mapping them. Quick and dirty work around until it is fixed.

![CleanShot 2024-10-29 at 15 08 04](https://github.com/user-attachments/assets/c31891d8-60c2-4824-9bf2-8dcd111d31bc)

**Comment by PierBover** [comment]
Same issue here. This started happening after updating to the latest version yesterday.

**Comment by alexpani** [comment]
Thanks, it works for me too in italiano keyboard.

**Comment by Dubrefjord** [comment]
Same issue on my macbook with a swedish keyboard. 

OS: sequoia 15.1.1
Karabiner version: 15.3.0

**Comment by tewzdk** [comment]
is still happening in v15.3.0 on Danish keyboard.

**Comment by bintoro** [comment]
Question to those suffering from this issue: is it really the case that the ANSI vs. ISO choice under the “Virtual Keyboard” section makes no difference?

I’m asking because, with my ANSI keyboard, the ANSI/ISO setting definitely switches the two keys around.

The USB usage code (which is what Karabiner outputs) for each physical key is independent of the keyboard layout or language. The key to the left of “1” always has the usage code `0x35`. The emitted usage codes can be inspected in Karabiner’s EventViewer app.

With an ANSI keyboard, `0x35` gets translated to the Mac keycode `0x32` by the OS. This keycode in turn gets translated to, e.g., the backtick character if the US keyboard layout is active.

With an ISO keyboard, `0x35` gets translated to the Mac keycode `0x0A` by the OS. This keycode in turn gets translated to, e.g., “§” if the Finnish/Swedish keyboard layout is active.

The key to the right of LShift, which is only found on ISO keyboards, has the USB usage code `0x64` but the same Mac keycode `0x32` as the backtick key on ANSI keyboards.

So, this issue sounds as if macOS is translating the USB codes to Mac keycodes according to the ANSI layout even though the physical device has an ISO layout. But this is exactly what the ANSI/ISO setting in Karabiner is supposed to control — and it does, at least for me. If I change the setting to ISO (with my ANSI keyboard), the backtick key begins to output `0xA0` as expected, and thus if I switch to a Finnish keyboard layout, I get “§” as expected.

**Comment by tewzdk** [comment]
@bintoro I just tried what you said and I changed my setting from ISO to ANSI, and now it works somehow.

I have M2 macbook pro with german keyboard, but I have the settings set to Danish.

Thanks even though it is not exactly the same as for you, it somehow works.

**Comment by bintoro** [comment]
@tewzdk ISO is the one that _should_ work for European layouts.

The key next to “1” should produce the Mac keycode `0x0A` (ISO) or `0x32` (ANSI). If these are switched, there’s something wrong.

The USB usage code (visible in EventViewer) should always be `0x35` for that key regardless of the physical layout or any settings. If it isn’t, then that would be very interesting information.

Given that the issue appeared in v15, it may have to do with this release note:

> To improve the behavior around the fn key, the Vendor ID and Product ID of the virtual keyboard used by Karabiner-Elements have been changed to the same values as those of a real Apple external keyboard.

I took a quick look at the code, but no luck. The reported Product IDs do seem to correctly account for the ANSI/ISO setting.

**Comment by bintoro** [comment]
A cursory Google search indicates that macOS confusing an ISO keyboard for an ANSI one is actually not uncommon irrespective of Karabiner. This issue pops up all over the place because it has been going on since forever and because there is no way to change the initially chosen type via the UI.

So, what may have happened is that, when Karabiner updated the reported USB Product ID, macOS for some reason mistakenly identified the new device as ANSI.

The information is stored in `/Library/Preferences/com.apple.keyboardtype.plist`. Apparently deleting this file and restarting if necessary will trigger the keyboard setup assistant that allows one to choose ISO vs. ANSI, but only if macOS cannot otherwise identify the type.

If that doesn’t help, I don’t see why one couldn’t just edit the file using `defaults` or `PlistBuddy`. Mine has a value of `40` at all entries, so I guess `40` corresponds to ANSI.

I’d be curious to hear what `defaults read /Library/Preferences/com.apple.keyboardtype.plist` prints for the ISO guys.

**Comment by PierBover** [comment]
Yeah for some reason macOS Ventura misidentified my current keyboard type as `ANSI` instead of `ISO`. I had to delete `/Library/Preferences/com.apple.keyboardtype.plist` to reset this and restart the process again. Even if I started the process using the GUI that didn't work (by going to System Settings, Keyboard, ).

> I’d be curious to hear what defaults read /Library/Preferences/com.apple.keyboardtype.plist prints for the ISO guys.

I have multiple keyboards. A mix of Spanish ISO and LATAM. Not sure why Apple Mexico stopped shipping their laptops with Spanish ISO a couple of years ago.

```
{
    keyboardtype =     {
        "257-9494-0" = 40;
        "303-9494-0" = 40;
        "45929-1133-0" = 41;
        "67-7247-0" = 41;
    };
}
```

**Comment by bintoro** [comment]
@PierBover’s comment appears to confirm that the issue is macOS misidentifying the keyboard type.

I suppose it would be worth investigating whether this now happens for all new Karabiner installations and if there is something Karabiner could be doing to help the keyboard get detected correctly. Beyond that, it’s not really a Karabiner bug.

To summarize, the possible solutions are:

**Set the keyboard type in Karabiner to ANSI.** Given that the cause is macOS thinking the keyboard is ANSI, specifying ANSI in Karabiner as well is actually a completely valid approach. There shouldn’t be any adverse effects.

**Set the correct  keyboard type on macOS’s end.** Either force keyboard redetection by deleting `/Library/Preferences/com.apple.keyboardtype.plist` or manually change the relevant entry from `40` to `41` by running

    sudo PlistBuddy -c "Set :keyboardtype:00000-0000-0 41" /Library/Preferences/com.apple.keyboardtype.plist

where `00000-0000-0` is the relevant device identifier.

**Comment by Chabulga** [comment]
I did both, first changing the entries from one 40 and the other from 43 to 41. I didn’t see any change of keyboard behavior.

After that I deleted the file and rebooted and the file was still gone and also no change in behavior.

**Comment by SnackForever** [comment]
> defaults read /Library/Preferences/com.apple.keyboardtype.plist

Worked for me thanks !

**Comment by e-gaulue** [comment]
Maybe the fact I don't even have a `/Library/Preferences/com.apple.keyboardtype.plist` is just KE virtual keyboard is not associated with an ISO, ANSI, JIS keyboard at all.

When I go back to 14.13.0, macos request me to press the key right to the left shift and this file exists with:
```
{
    keyboardtype =     {
        "10203-5824-0" = 43;
    };
}

```

**Comment by PierBover** [comment]
I've confirmed this is not a Karabiner issue.

I'm seeing it on a new Mac with macOS Sequoia where I never installed Karabiner. I'm using a Spanish ISO keyboard (a Keychron B1).

Unfortunately the trick of deleting `/Library/Preferences/com.apple.keyboardtype.plist` doesn't work anymore for me. Not sure if this only worked in Ventura or what.

**Comment by nejmeijer** [comment]
This issue seems related to #4295, is it really a different problem? If not, I will close one of them as a duplicate.

**Comment by Chabulga** [comment]
There are seval copies of this thread here that relate to the same problem. I know about #3987 , #4035 , #4102 , #4178  and  #4295 that you mentioned. For me the issue is still existing and the purposed midification of keyboardtype.plist did not solve the issue. I am still using the key swap rules that I posted here [Post](https://github.com/pqrs-org/Karabiner-Elements/issues/3984#issuecomment-2444386809)

**Comment by nejmeijer** [comment]
Thank you for linking those issues together! I guess this problem deserves more attention if it has been reported that many times.

**Comment by Chabulga** [comment]
Also #1842 seems to report this problem already back in 2019

**Comment by habibalamin** [comment]
I use my laptop keyboard and an external keyboard, but only have one entry in the `keyboardtype.plist` file.

It's a 43, but if 40 is ISO and 41 or ANSI or whichever way round it is, what's 43?

Anyway, I've deleted the file, I'll get round to fixing it properly at some point. For now, I've just manually created simple modification entries in Karabiner to swap the two keys around.

(The keys swapped for me are `§` and `` ` ``, and both my keyboards use a British layout, guv'.)
